### PR TITLE
Optimize binary operations with 1-bit on lhs and 1 const on rhs

### PR DESCRIPTION
**Pull request by MaxGraey**
`expr | 1`  -->  `1`
 `expr & 1`  -->  `expr`
`expr == 1` --> `expr`
`expr != 1` --> `!expr`
where `maxBits(expr) == 1` i.e booleans

For tests with 64-bit cases we required https://github.com/WebAssembly/binaryen/pull/2879 merge first

### CODE REVIEW

**Comment by MaxGraey** [comment]
Fuzzed:
```
Invocations so far:
   FuzzExec: 17382
   CompareVMs: 4322
   CheckDeterminism: 1500
   Wasm2JS: 4083
   Asyncify: 4277

ITERATION: 20263
```

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -1227,6 +1224,40 @@ struct OptimizeInstructions
...
+              return Builder(*getModule()).makeUnary(WrapInt64, binary->left);
+            }
+            case NeInt32: {
+              // bool(x) != 1  ==>  !bool(x)
+              return Builder(*getModule()).makeUnary(EqZInt32, binary->left);
+            }
+            case NeInt64: {
+              // i64(bool(x)) != 1  ==>  !i64(bool(x))
+              return Builder(*getModule()).makeUnary(EqZInt64, binary->left);
+            }
```
Comment (tlively) [inline src/passes/OptimizeInstructions.cpp:1256]: Could this use `Abstract`?

**Review by tlively — COMMENTED** [review]
Generally looks good 👍

File: src/passes/OptimizeInstructions.cpp [hunk 2]
```diff
@@ -1227,6 +1224,40 @@ struct OptimizeInstructions
...
+        // bool(x) | 1  ==>  1
+        // bool(x) & 1  ==>  bool(x)
+        // bool(x) == 1  ==>  bool(x)
+        // bool(x) != 1  ==>  !bool(x)
+        if (Bits::getMaxBits(binary->left, this) == 1) {
+          switch (binary->op) {
+            case OrInt32:
+            case OrInt64: {
+              // bool(x) | 1  ==>  1
+              return binary->right;
```
Comment (MaxGraey) [inline src/passes/OptimizeInstructions.cpp:1236]: Probably we should check side effects for `binary->left` here, wdyt?

Comment (tlively) [reply src/passes/OptimizeInstructions.cpp]: Ah right, yes checking for side effects is important.

**Comment by MaxGraey** [comment]
refuzzed:
```
Invocations so far:
   FuzzExec: 18569
   CompareVMs: 4691
   CheckDeterminism: 1553
   Wasm2JS: 4369
   Asyncify: 4535

ITERATION: 21610
```

File: src/passes/OptimizeInstructions.cpp [hunk 3]
```diff
@@ -258,9 +254,9 @@ struct OptimizeInstructions
             //    zero is impossible to be equal to a sign-extended value
             //    anyhow, so the entire thing is false.
             //  * if they were all one, we may get a false false, if the only
-            //    difference is in those upper bits. that means we are equal on
-            //    the other bits, including the sign bit. so we can just mask
-            //    off the upper bits in the constant value, in this case,
+            //    difference is in those upper bits. that means we are equal
+            //    on the other bits, including the sign bit. so we can just
+            //    mask off the upper bits in the constant value, in this case,
```
Comment (tlively) [inline src/passes/OptimizeInstructions.cpp:259]: There are still a few unrelated comment changes here and above that would be good to remove.

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp]: done!

File: test/passes/optimize-instructions_all-features.txt [hunk 4]
```diff
@@ -3417,14 +3417,73 @@
...
+  )
+  (drop
+   (i64.ne
+    (i64.and
+     (local.get $y)
+     (i64.const 1)
+    )
+    (i64.const 1)
+   )
+  )
```
Comment (tlively) [inline test/passes/optimize-instructions_all-features.txt:3459]: It looks like these two optimizations didn't trigger for some reason.

File: test/passes/optimize-instructions_all-features.txt [hunk 5]
```diff
@@ -3417,14 +3417,73 @@
...
+  (drop
+   (i64.or
+    (i64.and
+     (local.get $y)
+     (i64.const 1)
+    )
+    (i64.const 1)
+   )
+  )
```
Comment (tlively) [inline test/passes/optimize-instructions_all-features.txt:3486]: Same with these.

**Review by tlively — COMMENTED** [review]
Looks good now except that it looks like the i64 optimizations aren't working.

File: test/passes/optimize-instructions_all-features.txt (hunk 4 above)
Comment (MaxGraey) [reply test/passes/optimize-instructions_all-features.txt:3459]: Yes. It was expected. On description I mentioned this PR depend on this another: #2879

Comment (MaxGraey) [reply test/passes/optimize-instructions_all-features.txt:3459]: But #2879 pretty complex and should be double check. Not sure it ready yet

**Review by tlively — APPROVED** [review]
🎉
